### ISSUE TITLE
View request description history

### ISSUE THREAD

**Issue by lmullane** [bcgov/foi-flow#31]
As an IAO analyst or ministry coordinator, I want to be able to view the request description history so that I can better understand what the applicant is requesting.

**Assumption**	
We will retain the all changes to the request description AND/OR the date range from what was originally submitted by the applicant.

**Acceptance Criteria**	

**Scenario 1: View description history**
* GIVEN an analyst or ministry coordinator has saved an updated 'Request Description'
* WHEN the analyst is on the request view in any state
* THEN the 'Description History' link above the request description field will become active. 
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/768088423

**Scenario 2: Remove Personal Information Checkbox checked on Unopened Request**
* GIVEN an Intake Analyst is reviewing a request in Unopened state
* WHEN the analyst has checked the "Description contains NO Personal Information" checkbox
* AND has completed all other mandatory fields
* THEN the Save button will be enabled
* AND the request description will be saved as a new version known as the Redacted Request Description

** Scenario 3: Checkbox unchecked on Unopened Request**
* GIVEN an Intake Analyst is reviewing a request in Unopened state
* WHEN the analyst has NOT checked the "Description contains NO Personal Information" checkbox
* THEN the Save button will be disabled (greyed out)

**Scenario 4: Unable to remove checkbox once saved**
* GIVEN an IAO analyst is reviewing a request that has been Saved
* WHEN the analyst views the "Contains NO Personal Information" checkbox
* THEN the checkbox will be disabled (greyed out)
* AND will remain disabled in all subsequent states (Intake in Progress, Open, Call For Records, etc.)

**Scenario 5: IAO analyst view original request description**
* GIVEN the request description has been updated
* WHEN the IAO analyst clicks on 'Description History'
* THEN a modal will open with all the previous version of the request description history:
* https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/768088423
* AND the original request description will be view only (not editable) at the top with the original date range, original request description, 'Entered by applicant' or name of user (i.e. staff) who entered the original description
* AND the date the original description was entered

**Scenario 6: Show additional versions to IAO analysts of the description history**
* GIVEN the request description has been updated more than once
* WHEN the IAO analyst clicks on 'Description History'
* THEN a modal will open with all the previous version of the request description history:
* https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/768088423
* AND the below the original request description ordered from oldest to newest will be versions of the updated description history
* AND the only details visible will be the name of user (i.e. staff) who updated the description
* AND the date the description was updated

**Scenario 7: Show versions not including original to ministry coordinators of the description history**
* GIVEN the request description has been updated using the "Redact Personal Information" checkbox
* WHEN the ministry coordinator clicks on 'Description History'
* THEN a modal will open with all the previous version of the request description history starting from the Redact Personal Information version:
* https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/768088423
* AND the below ordered from oldest to newest will be versions of the updated description history (not including original)
* AND the only details visible will be the name of user (i.e. staff) who updated the description
* AND the date the description was updated

**Scenario 8: View details of additional versions of the description history**
* GIVEN the request description has been updated more than once
* AND the analyst or coordinator is on the 'Description History' modal
* https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/768088423
* WHEN the analyst or coordinator clicks on the 'arrow; to the right of the date stamp 
* THEN the details of the updated request description will open
* AND the details will be view only (not editable) with the date range, updated request description, user (i.e. staff) who made the updates, 
* AND the date the updates were entered

**Dependencies? 

How does editing a field work in Material UI?

As noted in the comments, OI staff should be able to edit the request description for requests that are 'Closed' (i.e. editing will be role specific).

**Validation Rules? (If yes, list here)**	

**Design**
@xxx - please link the Design here

**Definition of Ready**

1. [ ] Is there a well articulated User Story?
2. [ ] Is there Acceptance Criteria that covers all scenarios (happy/sad paths)?
3. [ ] If there is a user interface, is there a design? 
4. [ ] Does the user story need user research/validation?
5. [ ] Does this User Story needs stakeholder approval?
6. [ ] Design / Solution accepted by Product Owner
7. [ ] Is this user story small enough to be completed in a Sprint? Should it be split?
8. [ ] Are the dependencies known/ understood? (technical, business, regulatory/policy)
9. [ ] Has the story been estimated?

**Definition of Done**

1.     [ ] Passes developer unit tests
2.    [ ] Passes peer code review
3.    [ ] If there's a user interface, passes UX assurance
4.    [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
5.	[ ] Confirm Test cases built and succeeding 
6.	[ ] No regression test failures
7.	[ ] Test coverage acceptable by Product Owner
8.	[ ] Ticket ready to be merged to master or story branch
9.	[ ] Developer to list Config changes/ Update documents and designs
10.	[ ] Can be demoed in Sprint Review 
11.	[ ] Tagged as part of a Release
12.   [ ] Feature flagged if required
13.  [ ] Change Management activities done?

**Comment by lmullane** [comment]
@arielleandrews, @m-prodan , I assume that as an applicant can narrow their request after intake, then the analyst should be able to edit the request description no matter the status of the request (i.e. Unopened, Open, Call for Records, Fee Estimate, etc.) Exceptions would likely be 'Ministry Approval' and 'Closed.'

Similarly they could edit the date range of the request at any time/status, with the exception of a 'Ministry Approval' and 'Closed' request.

Does that sound correct?

**Comment by m-prodan** [comment]
@lmullane - that is correct, FOI analysts outside of intake will often change request descriptions if requests attract large fees. This includes both the description and date ranges.

**Comment by m-prodan** [comment]
@lmullane - a thought just occurred to me - we may need to be able to edit request descriptions after a request is closed. Because we publish request descriptions on generals, the Open Information team will always review request descriptions to ensure there are no potentially personal information in the request description, that wasn't anonymized by Intake. It could be a case of being able to re-open requests to do just that, or perhaps give specific permissions to allow OI analysts to do it without re-opening the request (would make their lives easier I'm sure if they could do it directly without going through steps of re-opening)

**Comment by arielleandrews** [comment]
I would say analysts need to be able to edit the request description at any stage, with the exception of Closed. Even at ministry approval things can change. Closed would be the only stage where analysts shouldn't be able to edit the description, with the exception for OI as Matt noted (although this scenario would be rare)

**Comment by lmullane** [comment]
Good feedback, @m-prodan and @arielleandrews. Will update the scenario that editing closed will be restricted to OI.

**Comment by lmullane** [comment]
@arielleandrews, @m-prodan, should ministry coordinators be able to edit a request description or date range, or should this be permitted only by IAO analyts?

**Comment by arielleandrews** [comment]
@lmullane no ministry coordinators should not be able to edit the request description or date range, you are correct this should be restricted to IAO staff

**Comment by lmullane** [comment]
Perfect. Thanks for the confirmation.

**Comment by lmullane** [comment]
@m-prodan, @arielleandrews, @JHarrietha-AOT, please review the scenarios/acceptance criteria for this story.

@JHarrietha-AOT, note I'm not sure how the analyst would be able to view the original request description but just put some filler in there until we have a design.

**Comment by arielleandrews** [comment]
@lmullane re the assumptions, a request description can change multiple times over the lifecycle of the FOI request, so we will need to save all edited versions. Intake is likely to only update the description once so this may be fine at this stage but will need to keep that in mind for future.

**Comment by JHarrietha-AOT** [comment]
@lmullane Adding the design for the user will view the previous description(s) is on my list to add today and will be reflected in the updated prototype I share out Monday. Do we want to include multiple previous versions as it can change multiple times over the lifecycle of the request, as @arielleandrews mentioned? Or just see the original description for now with the focus being intake?

**Comment by m-prodan** [comment]
Need to update this ticket to change some assumptions and consider renaming ticket. (past request descriptions and pulling archived changes when needed)

**Comment by JHarrietha-AOT** [comment]
@lmullane @m-prodan @arielleandrews Made some changes to the AC let me know what you think. I was thinking the analyst would want to mainly see the original request but would also want to see the iterative request descriptions and have access to see what changes were made and change back to any of the previous requests descriptions

**Scenario 1: Click Request Description field**

GIVEN the analyst is on an request that is an state other than 'Closed':
https://jacklyn808742.invisionapp.com/console/share/XU2L77IGEA/631957827
WHEN the analyst clicks in the 'Request Description' field
THEN the request description field will become editable

**Scenario 2: Save edits**

GIVEN the analyst has made edits to the 'Request Description'
WHEN the analyst clicks the 'Save' button
THEN the request description will be updated
AND the original request description will be archived

**Scenario 3: View archived description**

GIVEN an analyst has saved an updated 'Request Description'
WHEN the analyst is on the request view
https://jacklyn808742.invisionapp.com/console/share/XU2L77IGEA/631957827
THEN a 'View Request History' will appear next to the request description field
AND the original request description will be archived

**Scenario 4: View archived description**

GIVEN an analyst has saved an updated 'Request Description'
WHEN the analyst clicks 'View Request History'
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/724148981
THEN the Request History Pop Up appears
AND the archived request descriptions cannot be edited

**Scenario 5: Change back to archived request description**

GIVEN an analyst is on the Request History Pop-Up
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/724148981
WHEN the analyst clicks Update to Archived Description
THEN the pop-up will close
AND  the selected archived request description will show in the request description field
AND the description can be edited

**Comment by abin-aot** [comment]
@lmullane @arielleandrews 

- So, by the **Original Description**, I am assuming the **Description** submitted by the Applicant, Right?
 - Assuming bottom Accordions have the Descriptions(archived/versioned) one in a sort order? What is Sorting order?
- Does those accordion panel also have "update to archived description" button, so that analyst can take that version  of description for edit?


- Just out of curiosity, we are versioning all the metadata of the request including Description, Search dates, Ministries Selected, requestor details and many more etc. So, why "**Request History**" pop up ( as it  is named)  shows up only  Description ? . Does this meant only for "Request Description History" ?

- Do we need to indicate who updated it previously edit/entered those descriptions on the pop up? Also do we need the version number some where?

**Comment by arielleandrews** [comment]
- Original Description would be the Description submitted by the applicant on the webform yes
- Sort order - my preference would be newest to oldest
- Unsure of this one without seeing the updated design - I would think if the Analyst needs to change the description they could just edit it in the text box, which would already be showing the most recent version, and then once Saved that version would then be moved to Archived - can discuss in refinement
- Good point - the dates can change as part of the Description so we would need to capture and display this. Ministries may not be necessary as we would only have the multiple ministries captured on the initial raw request, once they are opened the requests are for one Ministry only so it won't change after that (and if it does it would mean closing and opening a new request)
- Displaying who updated the description would be helpful. Version number not a requirement in my mind but I'm curious to hear @m-prodan 's thoughts on that

**Comment by lmullane** [comment]
@mpilchar we will refine this ticket in Sprint Planning tomorrow:)

cc @divyav-aot, @sumathi-thirumani-aot, @richard-aot, @abin-aot, @arielleandrews, @JHarrietha-AOT

**Comment by arielleandrews** [comment]
@JHarrietha-AOT will update design to include a checkbox for the original request description "Remove Personal Information" to create a new version of the request description. The Ministry Coordinator persona will then see the checkbox onward version as the first version of the request description

**Comment by divyav-aot** [comment]
@arielleandrews - can intake and flex team member see original description before redaction?

**Comment by arielleandrews** [comment]
@divyav-aot yes any IAO staff can see the original request description

**Comment by arielleandrews** [comment]
@divyav-aot Scenarios have been updated in this ticket to reflect more detail on the No Personal Information checkbox. I will close #491 as duplicate. Jackie's updated design: 
![image](https://user-images.githubusercontent.com/82615839/133650027-0164789d-e8d0-48c4-baa0-1cff1db0ac71.png)

**Comment by lmullane** [comment]
Thanks, @arielleandrews. I was going to ask we make this change. You're on it.

**Comment by divyav-aot** [comment]
@JHarrietha-AOT , @arielleandrews  - one query regarding the username in description history.
If the original request is from OnlineForm then, we don't have a username for that. What should we display as username? "Online Form".
![image.png](https://images.zenhubusercontent.com/6081a4273965c56c02fdc82c/bb546b57-5daa-4de4-87df-fc4897e54166)

**Comment by arielleandrews** [comment]
@divyav-aot 'Online Form' makes sense to me.

**Comment by divyav-aot** [comment]
thanks @arielleandrews

**Comment by m-prodan** [comment]
I've completed my testing for this one - all seems to pass AC, but there is one thing I am confused by that I've not been able to test properly:

Scenario 7: Show versions not including original to ministry coordinators of the description history
GIVEN the request description has been updated using the "Redact Personal Information" checkbox
WHEN the ministry coordinator clicks on 'Description History'
THEN a modal will open with all the previous version of the request description history starting from the Redact Personal Information version:
https://jacklyn808742.invisionapp.com/console/share/3W2O64AP5C/768088423
AND the below ordered from oldest to newest will be versions of the updated description history (not including original)
AND the only details visible will be the name of user (i.e. staff) who updated the description
AND the date the description was updated

There is no redact personal information checkbox that I could see, and so the Ministry coordinator so far will always see the original request description. Was this scoped out at some point and the AC not updated? Thoughts @arielleandrews @divyav-aot @lmullane @mpilchar?

Everything else seems to be working as expected, I am going to update the AC for scenario 1, as I discussed yesterday, to reflect that description history will always be present.
